libmspack/mspack/mspack.h

```diff
@@ -506,6 +509,61 @@ struct mspack_file {
 /** Error code: error during decompression */
 #define MSPACK_ERR_DECRUNCH    (11)
 
+/**
+ * Returns a string with an error message appropriate for the last error
+ * of an mspack compressor or decompressor.
+ * 
+ * For use with mspack compressor and decompressor struct pointers.
+ *
+ * Example usage:
+ *
+ * @code
+ *   if (chmd->extract(chmd, f[i], outname)) {
+ *       printf("%s: extract error on \"%s\": %s\n",
+ *            *argv, f[i]->filename, MSPACK_ERROR(chmd));
+ *   }
+ * @endcode
+ *
+ * @param  base   An mspack compressor or decompressor struct pointer.
+ * @return a constant string with an appropriate error message.
+ */
+#define MSPACK_ERROR(base) mspack_error_msg(base->last_error(base))
+
+/**
+ * A function to convert the MSPACK error codes into strings.
+ *
+ * Example usage:
+ *
+ * @code
+ *   if (err != MSPACK_ERR_OK) {
+ *       fprintf(stderr, "%s -> %s: %s\n", argv[1], argv[2], mspack_error_msg(err));
+ *   }
+ * @endcode
+ *
+ * @param  int   An MSPACK_ERR code.
+ * @return a constant string with an appropriate error message.
+ */
+static inline const char *mspack_error_msg(int error) {
```

Comment:
Reviewing the code in light of these comments:

> Firstly, libmspack should be able to compile without _any_ standard C library functions when `MSPACK_NO_DEFAULT_SYSTEM` is defined. I understand that `static inline` means that technically this code is not in the library, but I would prefer to see no `<stdio.h>` in mspack.h at all.

... I've found that `memcmp`, `memset` and `strlen` calls have crept into libmspack, so I've introduced replacement functions in system.h and avoid including <string.h> at all, if MSPACK_NO_DEFAULT_SYSTEM is set, so libmspack still meets its std-C-library independence goals.

However, this makes it hard to include `system.h` with code outside libmspack when MSPACK_NO_DEFAULT_SYSTEM is set, so I moved the convenient macros that outside code was using `system.h` for into a new file, `macros.h`.

On reflection, I'd probably be happy with the error text function in mspack.h, if it simply returned "unknown error" rather than return a snprintf() buffer, and thus mspack.h did not need stdio.h

> Secondly, I am secretly trying to avoid any user-readable text available from the library API or header, because if it exists, it should really be internationalisable / localisable, therefore that would mean the library and/or header need to support every environment's localisation processes. I'd like to keep localisation exclusively to clients, so that e.g. clients depend on libmspack and gettext, not that libmspack depends on gettext, and definitely not that you need gettext installed in order to `#include <mspack.h>`

While I am trying to avoid it, it's clearly not working because there are several english-only warning messages emitted by libmspack. So this is still an issue to consider, this new macro doesn't *introduce* a new problem.

> Thirdly, it reduces the quality of cabextract. `cabextract.c` intentionally uses its own mspack_system implementation, and knows it uses C stdio functions, so its `error_msg()` function is not the same as `error.h`, it intentionally uses `errno` and `strerror()` to print a more useful error message that includes the reason _why_ there was a file opening/read/write error.

cabextract.c should still revert to its original implementation.

Reply:
Roger I'd be happy to revert cabextract to it's original error message implementation. I'll do this as I rebase the PR and fix merge conflicts. 

Regarding `mspack/macros.h`, I see that it's been used in one of the example programs.  Would you mind me removing the `#include` from `cabrip.c` and copypasting the code to define `LD`? This way, the example program only depends on header(s) provided by the `mspack` library build target.  

In the "modern" CMake method, a build is a collection of build targets to include objects, static or shared libraries, and executables. Each may have "private", "public", and "interface" sources.  
- Private sources are required to build the given target. 
- Public sources are required to build the given target and are provided as a part of the API for downstream targets. 
- Interface sources aren't used in the current target and provide an API for downstream targets (particularly useful when providing a header for pre-built sources, like a proprietary library. 

I've set up the examples as executable targets that link with the `mspack` shared library target.  As the `mspack` shared library doesn't include `mspack/macros.h` in it's public sources, the example programs don't have access to it, just like a real user application.  We could definitely cheat with the `target_include_directories()`, but it seems an appropriate test of the library to build the example executables the same way your library users would. 

As a side note, this also affects the tests targets, but for unit tests I think it makes sense to "cheat" and give unit tests access to library internals.

Reply:
The issue of only using exported headers is also the reason providing `MSPACK_ERROR()` and `mspack_error_msg()` in `mspack.h` made sense to me.  It makes these available for use in your examples.  

The alternatives are to:
(A) duplicate `error.h` in the examples directory.
(B) merely print error codes instead of error messages in the examples.  

Which would you prefer?  I lean towards duplicating error.h so users are guided towards the approach cabextract uses.  Since it's late and I want to push up updates before the work week, I'll make that change.  Let me know if you'd prefer something different.

Reply:
I think I would rather have error.h in the examples directory.

But in general, what I'd like to have with `macros.h` doesn't fit cleanly with CMake's way of thinking; a header file that is used both internally by a project, and internally by its examples, but simply because it helps define the coding conventions of that project, it's not part of that project's external interface, nor is it recommended for other project users.

Perhaps instead of copy-paste, would creating a symlink work? This is how cabextract consumes its internal copy of libmspack, and it also uses files "private" to libmspack in cabinfo, simply because it does not make sense (to me) to define the structure offsets of cabinet files more than once, even if this breaks the theoretical encapsulation.